src/core/algorithms/nar/des/des.cpp

```diff
@@ -0,0 +1,89 @@
+    RegisterOption(Option{&population_size_, kPopulationSize, kDPopulationSize, 100});
+    RegisterOption(
+            Option{&num_evaluations_, kMaxFitnessEvaluations, kDMaxFitnessEvaluations, 1000});
+    RegisterOption(Option{&differential_options_.differential_scale, kDifferentialScale,
+                          kDDifferentialScale, 0.5});
+    RegisterOption(Option{&differential_options_.crossover_probability, kCrossoverProbability,
+                          kDCrossoverProbability, 0.9});
+    RegisterOption(Option{&differential_options_.differential_strategy, kDifferentialStrategy,
+                          kDDifferentialStrategy, default_strategy});
+}
+
+void DES::MakeExecuteOptsAvailable() {
+    using namespace config::names;
+    MakeOptionsAvailable({kPopulationSize, kMaxFitnessEvaluations, kDifferentialScale,
+                          kCrossoverProbability, kDifferentialStrategy});
+}
+
+FeatureDomains DES::FindFeatureDomains(TypedRelation const* typed_relation) {
+    auto feature_domains = std::vector<std::shared_ptr<ValueRange>>();
+    for (size_t i = 0; i < typed_relation->GetNumColumns(); i++) {
+        std::shared_ptr<ValueRange> domain = CreateValueRange(typed_relation->GetColumnData(i));
+        feature_domains.emplace_back(domain);
+    }
+    return feature_domains;
+}
+
+std::vector<EncodedNAR> DES::GetRandomPopulationInDomains(FeatureDomains domains) const {
+    auto population = std::vector<EncodedNAR>();
+    for (int i = 0; i < population_size_; i++) {
+        population.emplace_back(EncodedNAR(domains, typed_relation_.get()));
+    }
+    auto compare_by_fitness = [](EncodedNAR const& a, EncodedNAR const& b) -> bool {
+        return a.GetQualities().fitness > b.GetQualities().fitness;
+    };
+    std::sort(population.begin(), population.end(), compare_by_fitness);
+    return population;
+}
+
+EncodedNAR DES::MutateIndividual(std::vector<EncodedNAR> population, size_t at) {
+    MutationFunction diff_func =
+            EnumToMutationStrategy(differential_options_.differential_strategy);
+    return (*diff_func)(population, at, differential_options_);
+}
+
+unsigned long long DES::ExecuteInternal() {
+    FeatureDomains feature_domains = FindFeatureDomains(typed_relation_.get());
+    std::vector<EncodedNAR> population = GetRandomPopulationInDomains(feature_domains);
+
+    for (int i = 0; i < num_evaluations_;
+         i++) {  // TODO: change num_evaluations type to something unsigned
+        size_t candidate_i = i % population_size_;
+        EncodedNAR mutant = MutateIndividual(population, candidate_i);
+        NAR mutant_decoded = mutant.SetQualities(feature_domains, typed_relation_.get());
+        double candidate_fitness = population[candidate_i].GetQualities().fitness;
+
+        if (mutant.GetQualities().fitness > candidate_fitness) {
+            population[candidate_i] = mutant;
+            nar_collection_.emplace_back(mutant_decoded);
```

Comment:
they are stack variables, shouldn't the compiler be able to figure out that its the last time we use them on its own and use `std::move`?

Reply:
> they are stack variables, shouldn't the compiler be able to figure out that its the last time we use them on its own and use `std::move`?

No. 
Consider the first line:
```cpp
population[candidate_i] = mutant;
```
The compiler need to decide which `operator=` of `EncodedNar` class to use. There are two overloads  to it: : one is move-assignment, the other is copy-assignment. Compiler generated them by default for your class, since `EncodedNar` class does not define a destructor and/or move constructor and/or copy constructor and/or .... etc. There are some rules to this. Nevermind, the fact is it generated those overloads:
```cpp
//copy assignment
EncodedNAR& operator=(const EncodedNAR& other);
//move assignment
EncodedNAR& operator=(EncodedNAR&& other) noexcept;
```

Reply:
Thank you for the in-depth reply